File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Power/MetallicizePower.ts
import GainBlockAction from "../Action/Common/GainBlockAction";
import { SoundMaster } from "../Audio/SoundMaster";
import { AtEndOfTurnPreEndTurnCardsMessage } from "../Events/AtEndOfTurnPreEndTurnCardsMessage";
import FileHelper from "../FileHelper";
import { LocalizationProperty } from "../Gen/DB/Localization";
import { StringHelper } from "../StringHelper";
import { AbstractPower, PowerType } from "./AbstractPower";


export class MetallicizePower extends AbstractPower{
    public static Id = "Metallicize";
    Icon: string = FileHelper.FormatPath("Powers/armor.png");
    Id: string = "Metallicize";
    Name: string = LocalizationProperty.Read("能力-金属化");
    get Type(): PowerType {
        return PowerType.Buff;
    }

    GetDescription(...args: any[]): string {
        return StringHelper.FormatColorString(LocalizationProperty.Read("能力-金属化描述")).format(this.Amount);
    }

    PlayApplyPowerSfx() {
        super.PlayApplyPowerSfx();
        SoundMaster.PlayVoice("POWER_METALLICIZE",0.05);
    }

    AtEndOfTurnPreEndTurnCards(msg: AtEndOfTurnPreEndTurnCardsMessage) {
        super.AtEndOfTurnPreEndTurnCards(msg);
        this.Flash();
        this.AddToBot(new GainBlockAction(this.Owner,this.Amount));
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Power/Buff/MetallicizePower.cs
using ModTheSpire.Effect.Other;
using ModTheSpire.Event.Ids;
using TCG.Config;

namespace ModTheSpire.Power.Buff
{
    public class MetallicizePower : AbstractPower
    {
        public override string Id { get; } = nameof(MetallicizePower);
        public override string Name { get; } = LocalizationProperty.Read("能力-金属化");
        public override string Icon { get; } = "Powers/armor.png";
        public override PowerType Type { get; } = PowerType.Buff;
        public override string GetDescription(object[] args)
        {
            return string.Format(LocalizationProperty.Read("能力-金属化描述"), this.Amount);
        }

        public override void PlayApplyPowerSfx()
        {
            base.PlayApplyPowerSfx();
            SoundMaster.PlayVoice(SoundMaster.POWER_METALLICIZE,0.05f);
        }

        protected override void AtEndOfTurnPreEndTurnCards(AtEndOfTurnPreEndTurnCardsMessage msg)
        {
            base.AtEndOfTurnPreEndTurnCards(msg);
            Flash();
            AddToBot(new GainBlockAction(this.Owner,this.Amount));
        }
    }
}